ADR/0015-integration-service-two-phase-architecture.md

```diff
@@ -0,0 +1,109 @@
+# 15. The Two-phase Architecture of the Integration Service
+
+Date Documented: 2023-02-09
+
+## Status
+
+In review
+
+## Context
+
+The Integration Service is in charge of running integration test pipelines by executing
+the Tekton pipeline for each user-defined IntegrationTestScenario.
+The main payloads that are being tested are Snapshots which contain references to all
+Components that belong to the Application, along with their images.
+
+One problem faced by the integration service is caused by the fact that testing an
+application with Tekton pipelines takes time. During the course of testing an application,
+multiple component builds can happen in quick succession, leading to a potential race
+condition between different Snapshots that are created for each of those builds.
+This would primarily manifest itself by two Snapshots getting promoted in quick
+succession but neither of them having the latest images contained within them.
+
+## Decision
+
+In order to protect against the race conditions, the integration service will leverage a
+two phase approach to testing. The phases will consist of a Component phase that will
+always be executed and the optional Composite phase which would come into play only when
+race conditions between component builds are detected.
+
+### Component phase
+
+When a single component image is built, the Integration Service tests the application
+by creating a Snapshot. All Components with their images from the Global Candidate List
+are included within the Snapshot and then the Component that was newly built is
+updated/overwritten to complete the Snapshot creation.
+
+After all test pipelines for the Snapshot finish successfully, the Integration service
+updates the Global Candidate List with the newly built Component image and promotes
+the Snapshot. This is done only if the rest of the Global Candidate List matches
```

Comment:
> This is done only if the rest of the Global Candidate List matches the rest of the tested Snapshot components.

I think this differs from how I thought it was supposed to work. I thought that at the end of the Component phase (assuming that it passes), the newly built image is _always_ promoted to the Global Candidate List (unless we're operating on a PR). That way, the Composite phase always starts operating on an ApplicationSnapshot that is always just "built from" the current state of the Global Candidate List - whatever that is.

One undesirable side-effect of the design as I understood it, was that "the build could get broken". Meaning, two components A and B could both pass the first Component testing phase, and both get added to the Global Candidates List. Component A could then pass the Composite testing phase, but Component B could _fail_ the Composite testing phase if it somehow interacts badly with Component A.

At this point, the Snapshot that includes the newly built Component A but _not_ the newly built Component B would be promoted to the lowest static Environment, but the Snapshot that includes both the newly built Component A and the newly built Component B would _not_ be promoted. After this point, all new builds would fail until the user removes the broken Component B from the global candidate list (we have no documented way how to do this) or until the user merges a fix for Component B.

Reply:
> One undesirable side-effect of the design as I understood it, was that "the build could get broken". 

In our early architecting days I thought it was OK for "dev" or lowest possible static environment to be broken. I actually forgot the analogue that was shared with us but it suggested hat dev AWLAYS got deployed to whether things were broken or not? If it was broken, developers submitted new commits that would fix the issue and continue rolling things forward.

Reply:
Ok, so the key takeaway here is that the Global Candidate List should always be updated with the newly built component when the Composite phase passes successfully?

Reply:
Apologies, corrected my statement, what I meant to say is: 
```
The Global Candidate List should always be updated with the newly built component when the _Component_ phase passes successfully
```

Reply:
ACK. I will retract my previous comment in light of that.

If the Global Candidate List is always updated on a completed (success/fail) Component snapshot, then this will not necessarily result in broken dev environments as the snapshot itself might indicate failure (and therefore block promotion). In this case, we would just be blocking promotion to the lowest environment until we have a passing Snapshot. I don't have an issue with that interface as long as we are suitably notifying users of the failure (I assume it would be a PaC notification back to GH on the commit).

Reply:
Ok, in that case we can remove this sentence so the design matches the expectations:

> This is done only if the rest of the Global Candidate List matches the rest of the tested Snapshot components.

We will make the update to our controller logic as part of [HACBS-1779](https://issues.redhat.com/browse/HACBS-1779) so the successful Component pipeline always updates the Global Candidate list.
@ralphbean @arewm @Dannyb48

Reply:
I updated the wording and made the matching changes to the diagrams, please take a look